### ISSUE TITLE
Identical #defines generated with test code

### ISSUE THREAD

**Issue by fmg-richard-pearman** [astand/c-coderdbc#18]
Using the test dbc file and generate code. Search  in the header file for ???_U7_TEST_1_ro_CovFactor
You'll see that two #defined have been generated!
This is a result of only using the inner name of the signal and not having the PGN name or other unique identifier

BO_ 1911 SIG_TEST_1: 8 EPS
 SG_ U7_TEST_1 : 55|7@0+ (1,-255) [-255|-128] ""  BCM  <===================== Creates the same #define!

BO_ 333 UTEST_2: 8 BCM
 SG_ U8_TEST_1 : 39|8@0+ (1,0) [0|255] ""  BMS
 SG_ U7_TEST_1 : 47|7@0+ (1,-255) [-255|-128] ""  BMS    <===================== Creates the same #define!

**Comment by astand** [comment]
Hi @fmg-richard-pearman, could you please attach dbc file which leads to this issue?

**Comment by fmg-richard-pearman** [comment]
Hi @astand, I used the dbc file in the /test folder called testdbc.dbc from this repo.
[testdb.dbc.zip](https://github.com/astand/c-coderdbc/files/9947434/testdb.dbc.zip)

**Comment by fmg-richard-pearman** [comment]
The following changes fixed the issue

```diff
-fwriter->Append(sigprt->PrintPhysicalToRaw(&s, fdesc->gen.DRVNAME));
+fwriter->Append(sigprt->PrintPhysicalToRaw(&s, fdesc->gen.DRVNAME, m.Name));
```
and 

```diff
diff --git a/src/codegen/c-sigprinter.cpp b/src/codegen/c-sigprinter.cpp
index 82cda5c..52bb430 100644
--- a/src/codegen/c-sigprinter.cpp
+++ b/src/codegen/c-sigprinter.cpp
@@ -50,23 +50,28 @@ void CSigPrinter::LoadMessage(const MessageDescriptor_t& message)
   sigs_expr.push_back(nexpr);
 }
 
-std::string CSigPrinter::PrintPhysicalToRaw(const SignalDescriptor_t* sig, const std::string& drvname)
+std::string CSigPrinter::PrintPhysicalToRaw(const SignalDescriptor_t* sig, const std::string& drvname , const std::string& message_name)
 {
+  // Create the #define with the unique signal name - construct string once
+  std::string define_and_signal_identifier = StrPrint("#define %s_%s_%s", drvname.c_str(), message_name.c_str(), sig->Name.c_str());
+
+  // Get point to define_and_signal_identifier - get pointer once
+  char const *define_and_signal_identifier_c_str = define_and_signal_identifier.c_str();
+
   std::string retstr = "";
 
   retstr = StrPrint("// signal: @%s\n", sig->Name.c_str());
 
   if (sig->IsDoubleSig)
   {
-    retstr += StrPrint("#define %s_%s_CovFactor (%f)\n", drvname.c_str(), sig->Name.c_str(), sig->Factor);
+    retstr += StrPrint("%s_CovFactor (%f)\n", define_and_signal_identifier_c_str, sig->Factor);
   }
   else
   {
-    retstr += StrPrint("#define %s_%s_CovFactor (%d)\n", drvname.c_str(), sig->Name.c_str(), (int32_t)sig->Factor);
+    retstr += StrPrint("%s_CovFactor (%d)\n", define_and_signal_identifier_c_str , (int32_t)sig->Factor);
   }
 
-  retstr += StrPrint("#define %s_%s_toS(x) ( (%s) ", drvname.c_str(), sig->Name.c_str(),
-      PrintType((uint8_t)sig->TypeRo).c_str());
+  retstr += StrPrint("%s_toS(x) ( (%s) ", define_and_signal_identifier_c_str, PrintType((uint8_t)sig->TypeRo).c_str());
 
   if (sig->IsDoubleSig)
   {
@@ -91,7 +96,7 @@ std::string CSigPrinter::PrintPhysicalToRaw(const SignalDescriptor_t* sig, const
     }
   }
 
-  retstr += StrPrint("#define %s_%s_fromS(x) ( ", drvname.c_str(), sig->Name.c_str());
+  retstr += StrPrint("%s_fromS(x) ( ", define_and_signal_identifier_c_str);
 
   if (sig->IsDoubleSig)
   {
```

**Comment by astand** [comment]
Hi, @fmg-richard-pearman  yeah, I see that problem but not sure that it must fixed by appending name.
From the dbc file side we cannot have two (or more) signals with same name and different configuration, and if more than one message use one signal they will have same fromS and toS
From the code side, yeah, it looks pretty "messy" but it works.
Do you use some static analyzer or other tool which complains about duplications? In my case I don't see any warnings as long as duplicated defines have same values

**Comment by fmg-richard-pearman** [comment]
Hi @astand, I don't use any special analyzer, however the code is compiled using a C++17 compiler with warnings as errors.  When I tried to compile the code I got compiler warnings  =>errors.

By prefixing all signal names with the message name that the signal belongs to ensures signals are unique (unless the DBC contains two signals with the same name - which I think could be considered an error).  Perhaps another method would be to use the message id (rather than name)?

I also encountered another issue related to the VAL (enums).  In a dbc that I was using, some values had been defined with the same name but different values.

```
// VALUE
VAL_ 123 X_ANGLE 32763 "RESERVED"  32764 "RESERVED"  32765 "RESERVED"  32766 "ERROR" 32767 "OTHER" ;
```

This caused issues as the #defines generated all had the same name.  I looked at cantools to see what output it generated in this situation.  It added the value to the name. i.e.

```
#define X_ANGLE_RESERVED_32763_CHOICE (32763)
#define X_ANGLE_RESERVED_32764_CHOICE (32764)
#define X_ANGLE_RESERVED_32765_CHOICE (32765)
#define X_ANGLE_ERROR_CHOICE (32766)
#define X_ANGLE_OTHER_CHOICE (32767)
```
Makeing the name unique.

I hope this helps?

**Comment by astand** [comment]
Hi @fmg-richard-pearman 
Yeah it is more or less clear. Another way to fix this problem is to forbid generating macro for the signal more than one time. Putting the name maybe better solution, but I am thinking about amount of changes in already existed code. 
Thanks for participation

**Comment by astand** [comment]
And regrading VAL case - now it is done in a way with protected macro wrappers:
`
#ifndef TestValTableID_UTEST_3_Udef
#define TestValTableID_UTEST_3_Udef (-1)
#endif

#ifndef TestValTableID_UTEST_3_Udef
#define TestValTableID_UTEST_3_Udef (6)
#endif

#ifndef TestValTableID_UTEST_3_Udef
#define TestValTableID_UTEST_3_Udef (5)
#endif

Yeah, that is not the same as in cantools, but at least should not bother compiler at all`

**Comment by astand** [comment]
Fixed in v2.7. No duplication anymore (name is still same)
